### PR TITLE
Feature add support for QT and new '--enable-qt' option

### PR DESCRIPTION
**Pull request by cariepointer**
Adding wolfSSL support for QT 5.12

### CODE REVIEW

**Comment by dgarske** [comment]
@cariepointer : Review so far on the QT branch is good. I haven't seen anything of concern (about 50% through). Also ran make check and scan-build without any issues.

Question: How do I setup QT with its tests? Is this documented somewhere?

File: src/ssl.c [hunk 1]
```diff
@@ -25567,6 +27929,8 @@ int wolfSSL_DSA_do_verify(const unsigned char* d, unsigned char* sig,
     ret = DsaVerify(d, sig, (DsaKey*)dsa->internal, dsacheck);
     if (ret != 0 || *dsacheck != 1) {
         WOLFSSL_MSG("DsaVerify failed");
+        printf("\n\nret is: %d\n", ret);
```
Comment (dgarske) [inline src/ssl.c]: We should avoid printf and use WOLFSL_MSG or wrap in some type `#ifdef`.

File: src/ssl.c [hunk 2]
```diff
@@ -26377,6 +28747,53 @@ int wolfSSL_EVP_PKEY_set1_RSA(WOLFSSL_EVP_PKEY *pkey, WOLFSSL_RSA *key)
...
+        }
+        else {
+            WOLFSSL_MSG("wc_RsaKeyToPublicDer failed");
+        }
+        XFREE(derBuf, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+        wolfSSL_EVP_PKEY_free(pkey);
+        return WOLFSSL_FAILURE;
+    }
+
+    pkey->pkey.ptr = (char*)XMALLOC(derSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (dgarske) [inline src/ssl.c]: We should have the heap hint available in `WOLFSSL_EVP_PKEY` or `WOLFSSL_RSA`. Let's make sure and pass it for XMALLOC/XFREE when we have it.

File: src/ssl.c [hunk 3]
```diff
@@ -26389,27 +28806,182 @@ int wolfSSL_EVP_PKEY_set1_RSA(WOLFSSL_EVP_PKEY *pkey, WOLFSSL_RSA *key)
...
+        wolfSSL_DSA_free(pkey->dsa);
+    }
+    pkey->dsa    = key;
+    pkey->ownDsa = 0; /* pkey does not own DSA */
+    pkey->type = EVP_PKEY_DSA;
+
+    /* 4 > size of pub, priv, p, q, g + ASN.1 additional information */
+    derMax = 4 * wolfSSL_BN_num_bytes(key->g) + AES_BLOCK_SIZE;
+
+    derBuf = (byte*)XMALLOC(derMax, NULL, DYNAMIC_TYPE_TMP_BUFFER);
```
Comment (dgarske) [inline src/ssl.c]: Please use heap hint if available for XMALLOC/XFREE.

File: src/ssl.c [hunk 4]
```diff
@@ -20623,17 +22519,30 @@ const char* wolfSSL_state_string_long(const WOLFSSL* ssl)
...
-    (void)key;
-    WOLFSSL_STUB("PEM_def_callback");
-    return 0;
+    WOLFSSL_ENTER("wolfSSL_PEM_def_callback");
+
+    /* We assume that the user passes a default password as userdata */
+    if (key) {
+        sz = (int)XSTRLEN((const char*)key);
+        sz = (sz > num) ? num : sz;
+        memcpy(name, key, sz);
```
Comment (dgarske) [inline src/ssl.c]: Please use XMEMCPY.

File: wolfssl/internal.h [hunk 5]
```diff
@@ -3353,20 +3364,32 @@ typedef struct Arrays {
...
+#define STACK_TYPE_X509    0
+#define STACK_TYPE_NAME    1
+#define STACK_TYPE_BIO     2
+#define STACK_TYPE_OBJ     3
+#define STACK_TYPE_STRING  4
+#define STACK_TYPE_CIPHER  5
+#define STACK_TYPE_ACCESS_DESCRIPTION 6
+
 struct WOLFSSL_STACK {
+    byte type;     /* Set to STACK_TYPE_*. identifies type stored in data.*/
```
Comment (toddouska) [inline wolfssl/internal.h]: Placing a one byte member at the top of the structure is not efficient.  Members should be ordered by biggest to smallest.

File: wolfssl/internal.h [hunk 6]
```diff
@@ -3439,11 +3463,15 @@ struct WOLFSSL_X509 {
...
+    byte*            authInfoOcsp;
+#if defined(OPENSSL_ALL) || defined(WOLFSSL_QT)
+    byte*            authInfoCaIssuer;
+    int              authInfoCaIssuerSz;
+#endif
     word32           pathLength;
     word16           keyUsage;
     int              CRLInfoSz;
-    int              authInfoSz;
+    int              authInfoOcspSz;
```
Comment (toddouska) [inline wolfssl/internal.h]: Why is Ocsp added to the naming?

File: wolfssl/openssl/pem.h [hunk 7]
```diff
@@ -45,15 +45,26 @@ int wolfSSL_PEM_write_bio_RSAPrivateKey(WOLFSSL_BIO* bio, WOLFSSL_RSA* rsa,
...
 WOLFSSL_RSA* wolfSSL_PEM_read_bio_RSAPrivateKey(WOLFSSL_BIO* bio,
                                                   WOLFSSL_RSA**,
                                                   pem_password_cb* cb,
                                                   void* arg);
+
+WOLFSSL_API
+int wolfSSL_PEM_write_bio_RSA_PUBKEY(WOLFSSL_BIO* bio, WOLFSSL_RSA* rsa);
+
+WOLFSSL_API
+WOLFSSL_RSA *wolfSSL_PEM_read_bio_RSA_PUBKEY(WOLFSSL_BIO* bio, \
```
Comment (toddouska) [inline wolfssl/openssl/pem.h]: Why is a string continuation character being used the end of lines?

File: src/internal.c [hunk 8]
```diff
@@ -16536,6 +16561,187 @@ const char* GetCipherNameInternal(const byte cipherSuite0, const byte cipherSuit
...
+            break;
+        }
+    }
+
+    /* Protocol that matches OpenSSL's protocol version. */
+     switch (minor) {
+         case TLSv1_MINOR :
+             protocol = "TLSv1";
+             break;
+         case TLSv1_2_MINOR :
```
Comment (toddouska) [inline src/internal.c]: No TLS 1.3 support?

File: wolfcrypt/src/asn.c [hunk 9]
```diff
@@ -2316,8 +2324,11 @@ int ToTraditionalInline_ex(const byte* input, word32* inOutIdx, word32 sz,
     }
 
     ret = GetOctetString(input, &idx, &length, sz);
-    if (ret < 0)
-        return ret;
+    if (ret < 0) {
+        /* Don't error out if return is error - some private keys do not expect
```
Comment (toddouska) [inline wolfcrypt/src/asn.c]: Maybe so, but BUFFER_E is a real error, and idx and length may not be valid values but they're used as if they are now.

File: wolfcrypt/src/asn.c [hunk 10]
```diff
@@ -4064,15 +4121,27 @@ int DsaPrivateKeyDecode(const byte* input, word32* inOutIdx, DsaKey* key,
...
-        GetInt(&key->x,  input, inOutIdx, inSz) < 0 )
-        return ASN_DH_KEY_E;
+    if (GetInt(&key->p, input, inOutIdx, inSz) < 0 ||
+        GetInt(&key->q, input, inOutIdx, inSz) < 0 ||
+        GetInt(&key->g, input, inOutIdx, inSz) < 0 ||
+        GetOctetString(input, inOutIdx, &length, inSz) < 0 ||
+        GetInt(&key->y, input, inOutIdx, inSz) < 0) {
+            ret = ASN_PARSE_E;
+    }
+    if (ret == ASN_PARSE_E) {
```
Comment (toddouska) [inline wolfcrypt/src/asn.c]: A comment should be added to explain why this path is necessary.

File: wolfcrypt/src/asn.c [hunk 11]
```diff
@@ -9074,17 +9478,26 @@ int PemToDer(const unsigned char* buff, long longSz, int type,
...
         #ifdef HAVE_ECC
-            *eccKey = (header == BEGIN_EC_PRIV) ? 1 : 0;
+            *keyFormat = (header == BEGIN_EC_PRIV) ? 1 : 0;
         #else
-            *eccKey = 0;
+            *keyFormat = 0;
+        #endif
+        #ifndef NO_DSA
+            if (dsaFlag)
+                *keyFormat = 2;
```
Comment (toddouska) [inline wolfcrypt/src/asn.c]: Where is what these numbers mean documented?

File: src/ssl.c [hunk 12]
```diff
@@ -7139,6 +7237,848 @@ int wolfSSL_check_private_key(const WOLFSSL* ssl)
...
+                    break;
+
+                default:
+                    WOLFSSL_MSG("Unknown extension type found, parsing OID");
+                    /* If the extension type is not recognized/supported,
+                        set the ASN1_OBJECT in the extension with the
+                        parsed oid for access in later function calls */
+
+                    /* Get OID from input */
+                    GetASNObjectId(input, &idx, &length, sz);
```
Comment (toddouska) [inline src/ssl.c]: return code must be checked here.

File: src/ssl.c [hunk 13]
```diff
@@ -7139,6 +7237,848 @@ int wolfSSL_check_private_key(const WOLFSSL* ssl)
...
+                    if (input[tmpIdx++] != ASN_OCTET_STRING) {
+                        WOLFSSL_MSG("Error decoding unknown extension data");
+                        wolfSSL_ASN1_STRING_free(str);
+                        wolfSSL_ASN1_OBJECT_free(ext->obj);
+                        wolfSSL_X509_EXTENSION_free(ext);
+                        FreeDecodedCert(&cert);
+                        return NULL;
+                    }
+
+                    GetLength(input, &tmpIdx, &length, sz);
```
Comment (toddouska) [inline src/ssl.c]: return code must be checked here.

File: src/ssl.c [hunk 14]
```diff
@@ -7139,6 +7237,848 @@ int wolfSSL_check_private_key(const WOLFSSL* ssl)
...
+            /* Copy pathlen and CA into BASIC_CONSTRAINTS from object */
+            bc->ca = object->ca;
+            if (object->pathlen->length > 0) {
+                bc->pathlen = wolfSSL_ASN1_INTEGER_new();
+                if (bc->pathlen == NULL) {
+                    WOLFSSL_MSG("Failed to malloc ASN1_INTEGER");
+                    XFREE(bc, NULL, DYNAMIC_TYPE_X509_EXT);
+                    return NULL;
+                }
+                XMEMCPY((WOLFSSL_ASN1_INTEGER*)bc->pathlen, object->pathlen,
```
Comment (toddouska) [inline src/ssl.c]: memcpy doesn't look right for a type that may contain a pointer (data).  A deep copy is needed instead.

File: src/ssl.c [hunk 15]
```diff
@@ -15255,6 +16316,100 @@ char* wolfSSL_X509_NAME_oneline(WOLFSSL_X509_NAME* name, char* in, int sz)
...
+        /* Get short name */
+        sn = wolfSSL_OBJ_nid2sn(entry->nid);
+        if (sn == NULL) {
+            WOLFSSL_MSG("OBJ_nid2sn failed");
+            return NULL;
+        }
+
+        /* Copy sn and name text to buffer */
+        if (i != count - 1) {
+            strSz = (int)XSTRLEN(sn) + nameSz + 4;
```
Comment (toddouska) [inline src/ssl.c]: What does 4 mean?

File: src/ssl.c [hunk 16]
```diff
@@ -15896,7 +17201,56 @@ void wolfSSL_sk_GENERAL_NAME_pop_free(WOLFSSL_STACK* sk,
...
+
+     name = (WOLFSSL_GENERAL_NAME*)XMALLOC(sizeof(WOLFSSL_GENERAL_NAME), \
+                                           NULL, DYNAMIC_TYPE_OPENSSL);
+    if (name == NULL) {
+        WOLFSSL_MSG("Error allocationg space for GENERAL_NAME structure.");
+        return NULL;
+    }
+    XMEMSET(name, 0, sizeof(WOLFSSL_GENERAL_NAME));
+
+    name->d.dNSName = wolfSSL_ASN1_STRING_new();
```
Comment (toddouska) [inline src/ssl.c]: What if these _new(s) fail?

File: src/ssl.c [hunk 17]
```diff
@@ -16256,18 +17620,113 @@ void wolfSSL_sk_ASN1_OBJECT_free(WOLF_STACK_OF(WOLFSSL_ASN1_OBJECT)* sk)
...
+    if (str == NULL) {
+        WOLFSSL_MSG("Memory Error");
+        return NULL;
+    }
+    XMEMCPY(str, (unsigned char*)s->data, s->length);
+
+    tmp = (char*)XMALLOC(tmpSz, NULL, DYNAMIC_TYPE_TMP_BUFFER);
+    if (tmp == NULL) {
+        WOLFSSL_MSG("Memory Error");
+        return NULL;
```
Comment (toddouska) [inline src/ssl.c]: Looks like str is leaked here.

File: src/ssl.c [hunk 18]
```diff
@@ -16484,6 +17943,89 @@ const char* wolfSSL_get_cipher_name_from_suite(const byte cipherSuite0,
...
+#if defined(WOLFSSL_QT) || defined(OPENSSL_ALL)
+
+/* Creates and returns a new WOLFSSL_CIPHER stack. */
+WOLFSSL_STACK* wolfSSL_sk_new_cipher(void)
+{
+    WOLFSSL_STACK* sk = NULL;
+    WOLFSSL_ENTER("wolfSSL_sk_new_cipher");
+
+    sk = wolfSSL_sk_new_null();
+    sk->type = STACK_TYPE_CIPHER;
```
Comment (toddouska) [inline src/ssl.c]: What is sk is NULL?

File: src/ssl.c [hunk 19]
```diff
@@ -17820,16 +19434,126 @@ WOLFSSL_EVP_PKEY* wolfSSL_X509_get_pubkey(WOLFSSL_X509* x509)
...
+    int i,j,k;
+    int strLen;
+    unsigned long offset;
+
+    WOLFSSL_ENTER("wolfSSL_CIPHER_description");
+
+    if (!cipher->cipherOffset)
+        return NULL;
+
+    offset = 2*(cipher->cipherOffset);
```
Comment (toddouska) [inline src/ssl.c]: What does 2 mean here?

**Comment by aaronjense** [comment]
Commits for PR review requests will be squashed for easier review once all requests are addressed.

**Comment by dgarske** [comment]
@cariepointer : Looks like this PR has lots of merge conflicts to resolve. Assigning over to you for now. You might wait until the WebRTC PR is also merged in before spending time resolving.

**Comment by cariepointer** [comment]
Opened #2654, closing this PR

**Pull request closed without merging**